site/src/modules/resources/AgentLogs.tsx

```diff
@@ -0,0 +1,276 @@
+          if (index < logs.length - 1) {
+            nextChangesSource =
+              getLogSource(logs[index + 1].source_id).id !== log.source_id;
+          }
+          // We don't want every line to repeat the icon, because
+          // that is ugly and repetitive. This removes the icon
+          // for subsequent lines of the same source and shows a
+          // line instead, visually indicating they are from the
+          // same source.
+          if (
+            index > 0 &&
+            getLogSource(logs[index - 1].source_id).id === log.source_id
+          ) {
+            icon = (
+              <div
+                css={{
+                  width: 14,
+                  height: 14,
+                  marginRight: 8,
+                  display: "flex",
+                  justifyContent: "center",
+                  position: "relative",
+                  flexShrink: 0,
+                }}
+              >
+                <div
+                  className="dashed-line"
+                  css={(theme) => ({
+                    height: nextChangesSource ? "50%" : "100%",
+                    width: 2,
+                    background: theme.experimental.l1.outline,
+                    borderRadius: 2,
+                  })}
+                />
+                {nextChangesSource && (
+                  <div
+                    className="dashed-line"
+                    css={(theme) => ({
+                      height: 2,
+                      width: "50%",
+                      top: "calc(50% - 2px)",
+                      left: "calc(50% - 1px)",
+                      background: theme.experimental.l1.outline,
+                      borderRadius: 2,
+                      position: "absolute",
+                    })}
+                  />
+                )}
+              </div>
+            );
+          }
+
+          return (
+            <LogLine
+              line={logs[index]}
+              number={index + 1}
+              maxNumber={logs.length}
+              style={style}
+              sourceIcon={
+                <Tooltip
+                  title={
+                    <>
+                      {logSource.display_name}
+                      {assignedIcon && (
+                        <i>
+                          <br />
+                          No icon specified!
+                        </i>
+                      )}
+                    </>
+                  }
+                >
+                  {icon}
+                </Tooltip>
+              }
+            />
+          );
+        }}
+      </List>
+    );
+  },
+);
+
+export const useAgentLogs = (
+  agentId: string,
+  options?: { enabled?: boolean; initialData?: LineWithID[] },
+) => {
+  const initialData = options?.initialData;
+  const enabled = options?.enabled === undefined ? true : options.enabled;
+  const [logs, setLogs] = useState<LineWithID[] | undefined>(initialData);
+  const socket = useRef<WebSocket | null>(null);
```

Comment:
I didn't change it but when it is not enabled anymore, how we could close the socket connection without keeping the ref?

Reply:
`useEffect` would take care of that since `enabled` is part of the dependency array

So, let's say that we go through three state changes:
1. Initial render - `agentId` is available, but `enabled` is `false`
2. Re-render - `enabled` flips to `true`, and `agentId` stays the same
3. Re-render - `agentId` changes, but `enabled` is still `true`

With (1), the effect would end early and not set anything up.

But then we hit case (2) and the dependency array gets invalidated. There's no cleanup function from the previous effect to run, so we a new version of the effect callback with fresh closure values, creating a socket connection that is localized to the callback itself. We then end by returning out a cleanup function that will close out this effect callback's specific socket connection

But then we hit case (3), and the dependency array gets invalidated again. We do have a cleanup function from the last effect, so we run it to close out (2)'s connection, and then call a new effect callback with updated closure values. We make our new socket connection (still localized to the effect callback), and return out a cleanup function for (3)'s connection

So we're able to clean everything up like before – the big difference is that each socket connection isn't being exposed outside to the whole component. There's no shared state – each time an effect callback runs, the callback itself gets its own connection and closes its own connection

Reply:
Ahh ok, I got it.